include/circt/Dialect/FIRRTL/FIRRTLTypes.h

```diff
@@ -216,16 +220,21 @@ class IntType : public FIRRTLBaseType, public WidthQualifiedTypeTrait<IntType> {
 public:
   using FIRRTLBaseType::FIRRTLBaseType;
 
-  /// Return an SIntType or UIntType with the specified signedness and width.
+  /// Return an SIntType or UIntType with the specified signedness, width, and
+  /// constness.
   static IntType get(MLIRContext *context, bool isSigned,
-                     int32_t widthOrSentinel = -1);
+                     int32_t widthOrSentinel = -1, bool isConst = false);
 
   bool isSigned() { return isa<SIntType>(); }
   bool isUnsigned() { return isa<UIntType>(); }
 
   /// Return the width of this type, or -1 if it has none specified.
   int32_t getWidthOrSentinel();
 
+  /// Returns true if this is a 'const' type that can only hold compile-time
+  /// constant values
+  bool isConst();
```

Comment:
Wouldn't this just be a property of the base firrtl type?

Reply:
This overload of `isConst` on `IntType` is analogous to `getWidthOrSentinel` in that it dispatches to concrete subclasses.

In terms of only needing to define it once as a property of the base type, right now I don't believe there is a common base storage class for the FIRRTL types. If there's a straightforward way to do storage inheritance, I can look into that.

Reply:
I think I can make it work if I define all storage classes in C++, which is annoying but not that bad. Ironically [we just removed](https://github.com/llvm/circt/commit/b6d8508ba104b3120965dc031a3edfd538fb4602#diff-6fc1610e9971c8db86b15900035674d294fb7b694e34047749d2964987344088L724) some C++ type storage classes

Reply:
Alright, I've updated it to be stored in the base class.

Reply:
It should now be part of recursive types (hasConst?).

Reply:
What do you mean exactly? At least for the base int types, `isConst` needs to be stored explicitly since they don't store recursive type properties. 

If you just mean also adding a recursive `hasConst` property, that doesn't necessarily seem very useful except for maybe speeding up type comparisons (where, if two type pointers are not equal, we'd know the types are not equivalent if neither `hasConst`)